### ISSUE TITLE
A `$ref` to a fragment defined with `$id` in the same document isn't found.

### ISSUE THREAD

**Issue by rbuckton** [ajv-validator/ajv#1711]
**What version of Ajv are you using? Does the issue happen if you use the latest version?**
Version: 8.6.2

**Ajv options object**

```javascript
new Ajv({
  allErrors: true
  schemas: { "yaml.json": schema } // see below
});
```

**JSON Schema**

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "SeeAlso": { "$id": "#SeeAlso", "type": "number" },
    "Engine": {
      "$id": "#Engine",
      "type": "object",
      "properties": {
        "see_also": { "$ref": "#SeeAlso" }
      }
    }
  }
}
```

**Sample data**

```json
{
  "see_also": 1
}
```

**Your code**

```javascript
var Ajv = require('ajv');
var schema = {
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "SeeAlso": { "$id": "#SeeAlso", "type": "number" },
    "Engine": {
      "$id": "#Engine",
      "type": "object",
      "properties": {
        "see_also": { "$ref": "#SeeAlso" }
      }
    }
  }
};

var data = {
  see_also: 1
};

ajv = new Ajv({
    allErrors: true,
    schemas: {
      "yaml.json": schema
    }
});

var result = ajv.validate("yaml.json#/definitions/Engine", data);
console.log(result);
```

[View on runkit](https://runkit.com/rbuckton/60f9d3906738f8001aab8ca5)

**Validation result, data AFTER validation, error messages**

```
MissingRefError: can't resolve reference #SeeAlso from id yaml.json#Engine
```

**What results did you expect?**

As far as I can tell, it should be able to resolve fragment `#SeeAlso`, per https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.2.4 and https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.3.2.

**Are you going to resolve the issue?**

If I have some time I can put up a PR. If anyone else wants to take this on, my findings are below based on debugging ajv:

When the `$ref` keyword is evaluated for `{ "$ref": "#SeeAlso" }`, I end up here in _ref.js_:

![image](https://user-images.githubusercontent.com/3902892/126706975-cadc947b-94c7-445d-96f4-dba4cef82579.png "vocabularies/core/ref.js, line 19")

At this point, `baseId` is `yaml.json#Engine` and `$ref` is `#SeeAlso`.

When I step into `resolveRef`, the call to `resolveUrl` replaces `#SeeAlso` with `yaml.json#SeeAlso`:

![image](https://user-images.githubusercontent.com/3902892/126707282-1b3cf611-d0d2-402d-9f80-633b05d275c2.png "compile/index.js, line 130")

Since this isn't a root ref, I step through to line 136, where you might expect to resolve `#SeeAlso` against `root.localRefs`, however that fails because the fragment is qualified:

![image](https://user-images.githubusercontent.com/3902892/126707564-2176c1d7-9b86-4857-b423-a943ee93b6de.png "compile.index.js, line 136")

As you can see in the watch window, `root.localRefs` contains a definition for `#SeeAlso`, however the local ref can't be resolved because `ref` was changed from `"#SeeAlso"` to `"yaml.json#SeeAlso"`.

If this is *supposed* to work, then I expect the solution would be to check just the fragment (`#SeeAlso`) against `root.localRefs` (after ensuring that `ref` and `baseId` differ only by URL fragment).

I've run into this issue on and off for several years and have always worked around it by using `#/definitions/Foo`, but from what I can tell from reading the spec, this kind of `$ref` *should* work:

![image](https://user-images.githubusercontent.com/3902892/126708561-8daa68d4-9a3d-41b4-ae17-ebdeaab23d7e.png "json schema draft-07, example from page 15")

**Comment by rbuckton** [comment]
Apparently this only fails if the schema has no `$id`.

**Comment by epoberezkin** [comment]
correct, while there is no difference between schema key (the one you define in Ajv settings as you did), and schema $id for some contexts, there is a difference from the ref resolution point of view.

It does resolve references correctly if the schema has $id: https://runkit.com/esp/61065f45b2678e001953bec9

**Comment by rbuckton** [comment]
This still seems like a bug, since it doesn't follow the json-schema [Draft 7](https://json-schema.org/specification-links.html#draft-7) [core specification](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01).

The initial base URI is the URI at which the file is found, which in this case is the schema key, per [8.1](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.1) (**emphasis** mine):

> ### [8.1](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.1). Initial Base URI
> [RFC3986 Section 5.1](https://datatracker.ietf.org/doc/html/rfc3986#section-5.1) \[[RFC3986](https://datatracker.ietf.org/doc/html/rfc3986)\] defines how to determine the default base URI of a document.
>
> Informatively, **the initial base URI of a schema is the URI at which it was found**, or a suitable substitute URI if none is known.


An `$id` *overrides* the base URI, per [8.2](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.2) (**emphasis** mine):

> ### [8.2](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.2). The "$id" Keyword
> The "$id" keyword **defines a URI for the schema, and the base URI** that other URI references within the schema are resolved against.  A subschema's "$id" is resolved against the base URI of its parent schema.  **If no parent sets an explicit base with "$id", the base URI is that of the entire document**, as determined per [RFC3986 Section 5](https://datatracker.ietf.org/doc/html/rfc3986#section-5) \[[RFC3986](https://datatracker.ietf.org/doc/html/rfc3986)\]. \[...\]

And `$ref` references are resolved against the *base URI*, per [8.3](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.3) (**emphasis** mine):

> ### [8.3](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.3).  Schema References With "$ref"
>
> \[...\] The value of the "$ref" property MUST be a URI Reference.  **Resolved against the current URI base**, it identifies the URI of a schema to use.  \[...\]

It seems like its *close* to something that should work when looking at it in the debugger, since all of the fragment references are qualified based on the schema key.

**Comment by rbuckton** [comment]
If you're willing to accept a fix for this, it seems fairly straightforward:
1. In *core.ts* modify `Ajv._addSchema` to pass `baseId` into `getSchemaRefs.call`
2. In *resolve.ts* modify `getSchemaRefs` to accept `baseId` and pass it as a fallback to `normalizeId` (which you do pretty much everywhere else).

I can have a PR up with a test shortly, though building `master` and running `npm run test-spec` locally (*without* my changes) results in 156 test failures currently (run via Ubuntu using WSL on Windows).
